### PR TITLE
feat(experimental): support strict execution order

### PR DESCRIPTION
**Pull request by hyf0**
### Description

This feature try to solve the execution order problem caused by rollup and esbuild's code splitting logic. The nature of these problem is that the execution order of shared modules are hoisted unexpectedly.

- https://github.com/evanw/esbuild/issues/399
- https://github.com/evanw/esbuild/issues/2598
- https://github.com/vitejs/vite/issues/5142

This is done by adding helper function to simulate the original the semantic esm input while keep the static linking between modules.

### CODE REVIEW

**Comment by hyf0** [comment]
* **#1568** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/1568?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1567** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/1567?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @hyf0 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by Boshen** [comment]
May I suggest linking some prior arts or articles on what this is?

File: crates/rolldown/tests/fixtures/function/experimental/strict_execution_order/esbuild_issue_399/strict/init-dep-2.js
Comment (IWANABETHATGUY) [inline crates/rolldown/tests/fixtures/function/experimental/strict_execution_order/esbuild_issue_399/strict/init-dep-2.js:2]: Can you mock the function and do some assertion? since `console.log` is useless in unit testing, https://sevic.dev/notes/spies-mocking-node-test-runner/, I just found node has builtin support.

Comment (hyf0) [reply crates/rolldown/tests/fixtures/function/experimental/strict_execution_order/esbuild_issue_399/strict/init-dep-2.js:2]: I was going to, but I don't want change the original input that copied from issues. These tests are just examples to show output. We'll add more serious tests once the feature is going be landed stablely.

Comment (IWANABETHATGUY) [reply crates/rolldown/tests/fixtures/function/experimental/strict_execution_order/esbuild_issue_399/strict/init-dep-2.js:2]: yeah, just wanna this feature to be stable, you can go ahead and improve in the following pr.

File: crates/rolldown/src/stages/link_stage/wrapping.rs
Comment (IWANABETHATGUY) [inline crates/rolldown/src/stages/link_stage/wrapping.rs]: looks like a simple `matches!` would be more concise.

**Review by IWANABETHATGUY — DISMISSED** [review]
Looks good, maybe we also need to add some benchmark to compare the runtime performance between enabling `strict_execution_order` and disabling `strict_execution_order`

**Comment by hyf0** [comment]
> Looks good, maybe we also need to add some benchmark to compare the runtime performance between enabling `strict_execution_order` and disabling `strict_execution_order`

Yeah. Let's wait for we are sure that this feature is going to be landed in production.

**Comment by Boshen** [comment]
Do we have runtime tests to ensure the strictness? I remember webpack has them by asserting a whole bunch of things.

File: crates/rolldown_common/src/inner_bundler_options/types/experimental_options.rs
Comment (Boshen) [inline crates/rolldown_common/src/inner_bundler_options/types/experimental_options.rs:1]: This can go into `main` even if this PR doesn't get merged.

**Comment by Boshen** [comment]
For future reference, @hyf0 verified the reproduction https://github.com/vitejs/vite/issues/5142#issuecomment-953890608 is fixed with the feature in this PR.

**Pull request merged**
